lib/stdlib/src/uri_string.erl

```diff
@@ -0,0 +1,2067 @@
+%%    "6B%3C%" + "x"
+%% 7. Reverse Acc and return it
+%%-------------------------------------------------------------------------
+transcode([$%,_C0,_C1|_Rest] = L, Acc, InEnc, OutEnc) ->
+    transcode_pct(L, Acc, <<>>, InEnc, OutEnc);
+transcode([_C|_Rest] = L, Acc, InEnc, OutEnc) ->
+    transcode(L, Acc, [], InEnc, OutEnc).
+%%
+transcode([$%,_C0,_C1|_Rest] = L, Acc, List, InEncoding, OutEncoding) ->
+    transcode_pct(L, List ++ Acc, <<>>, InEncoding, OutEncoding);
+transcode([C|Rest], Acc, List, InEncoding, OutEncoding) ->
+    transcode(Rest, Acc, [C|List], InEncoding, OutEncoding);
+transcode([], Acc, List, _InEncoding, _OutEncoding) ->
+    lists:reverse(List ++ Acc).
+
+
+%% Transcode percent-encoded segment
+transcode_pct([$%,C0,C1|Rest] = L, Acc, B, InEncoding, OutEncoding) ->
+    case is_hex_digit(C0) andalso is_hex_digit(C1) of
+        true ->
+            Int = ?HEX2DEC(C0)*16+?HEX2DEC(C1),
+            transcode_pct(Rest, Acc, <<B/binary, Int>>, InEncoding, OutEncoding);
+        false -> throw({error, invalid_percent_encoding,L})
+    end;
+transcode_pct([_C|_Rest] = L, Acc, B, InEncoding, OutEncoding) ->
+    OutBinary = convert_binary(B, InEncoding, OutEncoding),
+    PctEncUtf8 = percent_encode_segment(OutBinary),
+    Out = lists:reverse(convert_list(PctEncUtf8, utf8)),
+    transcode(L, Out ++ Acc, [], InEncoding, OutEncoding);
+transcode_pct([], Acc, B, InEncoding, OutEncoding) ->
+    OutBinary = convert_binary(B, InEncoding, OutEncoding),
+    PctEncUtf8 = percent_encode_segment(OutBinary),
+    Out = convert_list(PctEncUtf8, utf8),
+    lists:reverse(Acc) ++ Out.
+
+
+%% Convert to binary
+convert_binary(Binary, InEncoding, OutEncoding) ->
+    case unicode:characters_to_binary(Binary, InEncoding, OutEncoding) of
+        {error, _List, RestData} ->
+            throw({error, invalid_input, RestData});
+        {incomplete, _List, RestData} ->
+            throw({error, invalid_input, RestData});
+        Result ->
+            Result
+    end.
+
+
+%% Convert to list
+convert_list(Binary, InEncoding) ->
+    case unicode:characters_to_list(Binary, InEncoding) of
+        {error, _List, RestData} ->
+            throw({error, invalid_input, RestData});
+        {incomplete, _List, RestData} ->
+            throw({error, invalid_input, RestData});
+        Result ->
+            Result
+    end.
+
+
+%% Flatten input list
+flatten_list([], _) ->
+    [];
+flatten_list(L, InEnc) ->
+    flatten_list(L, InEnc, []).
+%%
+flatten_list([H|T], InEnc, Acc) when is_binary(H) ->
+    L = convert_list(H, InEnc),
+    flatten_list(T, InEnc, lists:reverse(L) ++ Acc);
+flatten_list([H|T], InEnc, Acc) when is_list(H) ->
+    flatten_list(H ++ T, InEnc, Acc);
+flatten_list([H|T], InEnc, Acc) ->
+    flatten_list(T, InEnc, [H|Acc]);
+flatten_list([], _InEnc, Acc) ->
+    lists:reverse(Acc);
+flatten_list(Arg, _, _) ->
+    throw({error, invalid_input, Arg}).
+
+
```

Comment:
Maybe you want to use lists:flatmap!?

Reply:
I've measured 100% slowdown with the following flatmap/2 based implementation so I've kept the original function.

flatten_list([], _) ->
    [];
flatten_list(L, InEnc) ->
    Fun = fun(E) when is_binary(E) ->
                  convert_list(E, InEnc);
             (E) when is_list(E) ->
                  flatten_list(E, InEnc);
             (E) ->
                  [E]
          end,
    lists:flatmap(Fun, L).